src/Access/LDAPAccessStorage.cpp

```diff
@@ -128,7 +128,7 @@ void LDAPAccessStorage::processRoleChange(const UUID & id, const AccessEntityPtr
     {
         if (it != granted_role_names.end()) // Removed a granted role.
         {
```

Comment:
A comment with an explanation would be nice. The problem is that it->second is a reference to an object held by `granted_role_names`. Then in applyRoleChangeNoLock we first remove the object from `granted_role_names`, then try to remove the name from `granted_role_ids` but we are using the reference that has been just destroyed.

This problem also happens a few lines above with RENAME (L118), which should be adjusted too.

It seems this API is really error prone. Could we improve it instead?

Reply:
I mean the function definition (`void LDAPAccessStorage::applyRoleChangeNoLock(bool grant, const UUID & role_id, const String & role_name)`) and how it's used to remove roles. You need to know exactly how things are destroyed and how to manage the lifetime of the iterator, which seems suboptimal.